### DISCUSSION TITLE
SocketIO stops after Thread call

### DISCUSSION THREAD

**Discussion by wduandy** [miguelgrinberg/Flask-SocketIO#1802] [Q&A]
Hi, i'm using Flask-SockeIO and recently i've been implementing Threading Tasks. I'm getting a problem after the threading call, the Flask-SocketIO doesn't respond anymore. The threading call doesn't have anything about socket. I'm just using a app_context().

See the following code and how i implement it the thread call:

```python
from threading import Thread
from flask import render_template
from flask_mail import Message
from app import create_app, mail

app = create_app()

def send_async_email(app, msg):
  with app.app_context():
      mail.send(msg)

def send_email(subject, sender, recipients, text_body, html_body):
  msg = Message(subject, sender=sender, recipients=recipients)
  msg.body = text_body
  msg.html = html_body
  Thread(target=send_async_email, args=(app, msg)).start()

def send_password_reset_email(tutor):
  new_password = tutor.reset_password()
  send_email('Recuperação de Senha',
              sender=('admin', 'noreply@duvi.com'), 
              recipients=[tutor.usuario],
              text_body=render_template('email/inline/recuperar_senha.txt', tutor=tutor, new_password=new_password),
              html_body=render_template('email/inline/recuperar_senha.html', tutor=tutor, new_password=new_password))
```

I dont get any erros on terminal and i can't figure it out how to debug or solve the problem.

**Comment by miguelgrinberg** [comment]
Please don't write issues on this project when you only have a question about why your application does not work. The issues board is to report bugs. The discussions board is for your questions. Please follow the instructions next time.

What async mode are you using for your Socket.IO server? That is what's going to determine if threading is supported or not. If you want this to work, then the `threading` async mode is safe to use. If you are using eventlet or gevent, then you may need to apply their monkey patching to make threads, sockets and other parts of the Python library compatible.

**Comment by wduandy** [comment]
@miguelgrinberg, sorry for that! It is my first time writing something here.

For some reason i got to solve the issue by following your tutorial on 
https://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-xv-a-better-application-structure

Changing

`from app import create_app`

to

`from flask import current_app`

And in send_mail():

`Thread(target=send_async_email, args=(current_app._get_current_object(), msg)).start()`

Now i can't understand why this is working. Do you have any idea or need more info about the project?

Thanks and sorry one more time.

**Reply by miguelgrinberg** [reply]
Well, you didn't answer my questions. As I said above, threading working or not depends on the async mode that you are using.

The way you were using `app` in the non-working code was probably incorrect, because you were creating a different app instance for the email thread, instead of using the main one.
